File: python_test/algorithm/direct_predecessor_analyzer.py
"""
Phase 4 Prep: Direct Predecessor Analyzer

Analyzes direct predecessors (distance = 1 only) for vertical alignment.
"""

from typing import List, Tuple, Dict, Set


class DirectPredecessorAnalyzer:
    """
    Analyzes direct predecessors (distance = 1 only)
    """

    @staticmethod
    def compute_direct_predecessors(
        layers: List[List[str]],
        relations: List[Tuple[str, str]]
    ) -> Dict[str, Set[str]]:
        """
        Compute direct predecessors for each entity

        Direct predecessor: entity at distance exactly 1 (immediate neighbor)

        Returns: {entity: {direct_predecessors}}
        """
        # Build layer index
        layer_of = {}
        for layer_idx, layer in enumerate(layers):
            for entity in layer:
                layer_of[entity] = layer_idx

        # Find direct predecessors (layer distance = 1)
        direct_predecessors = {}
        for entity in layer_of:
            direct_predecessors[entity] = set()

        for left, right in relations:
            if left in layer_of and right in layer_of:
                left_layer = layer_of[left]
                right_layer = layer_of[right]

                # Direct if distance = 1 layer
                if right_layer - left_layer == 1:
                    direct_predecessors[right].add(left)

        return direct_predecessors

File: src/domain/services/layout/DirectPredecessorAnalyzer.ts
/**
 * Phase 4 Prep: Direct Predecessor Analyzer
 *
 * Analyzes direct predecessors (distance = 1 only) for vertical alignment.
 */

import { DirectedRelation } from './types';

export class DirectPredecessorAnalyzer {
  /**
   * Compute direct predecessors for each entity
   *
   * Direct predecessor: entity at distance exactly 1 (immediate neighbor)
   * i.e., entities in adjacent layers (layer_distance = 1)
   *
   * @param layers - Horizontal layer assignments
   * @param relations - Directed relations
   * @returns Map of entity -> set of direct predecessors
   */
  static computeDirectPredecessors(
    layers: string[][],
    relations: DirectedRelation[]
  ): Map<string, Set<string>> {
    // Build layer index
    const layerOf = new Map<string, number>();
    layers.forEach((layer, idx) => {
      layer.forEach(entity => {
        layerOf.set(entity, idx);
      });
    });

    // Find direct predecessors (layer distance = 1)
    const directPredecessors = new Map<string, Set<string>>();

    // Initialize for all entities
    layerOf.forEach((_, entity) => {
      directPredecessors.set(entity, new Set());
    });

    for (const { left, right } of relations) {
      const leftLayer = layerOf.get(left);
      const rightLayer = layerOf.get(right);

      if (leftLayer !== undefined && rightLayer !== undefined) {
        // Direct if distance = 1 layer
        if (rightLayer - leftLayer === 1) {
          directPredecessors.get(right)!.add(left);
        }
      }
    }

    return directPredecessors;
  }
}
